File: CAwaPI/sources/globalfactory.cpp
#include "../includes/globalfactory.h"

#include "../includes/hash.h"

#include <string.h>

static HashTable idPool;
static HashTable typePool;

List* components = NULL;

void GlobalFactory::init()
{
    hashTableInit(&idPool, 10);
    hashTableInit(&typePool, 10);
}

Component* GlobalFactory::addComponent(char* id, char* type, bool standardComponent)
{
    Component* obj = (Component*)g_factory.construct(type, id);
    
    hashTableInsert(&idPool, hash32(id, (int)strlen(id)), obj);
    
    List *l = (List*)hashTableLookup(&typePool, hash32(id, (int)strlen(id)));
    if (l == NULL)
    {
        l = (List*)malloc(sizeof(List));
        listInit(l);
        hashTableInsert(&typePool, hash32(id, (int)strlen(id)), l);
    }
    listAddElement(l, obj);
		
    return obj;
}

List* GlobalFactory::getComponentsByType(char* type)
{
    List* list = (List*)hashTableLookup(&typePool, hash32(type, (int)strlen(type)));
    if (list == NULL)
    {
        char* message = (char*)"No component typed ";
        message = stringInsert(message, type, (int)strlen(message));
        message = stringInsert(message, "exists for now", (int)strlen(message));
        
        throw message;
    }
    return list;
}

Component* GlobalFactory::getComponentById(char* id)
{
    Component* component = (Component*)hashTableLookup(&idPool, hash32(id, (int)strlen(id)));
    if (component == NULL)
    {
        char* message = (char*)"The component ";
        message = stringInsert(message, id, (int)strlen(message));
        message = stringInsert(message, " doesn't exist", (int)strlen(message));
        
        throw message;
    }
    return component;
}

File: JAwaPI/src/com/dviance/AwaPI/generic/GlobalFactory.java
package com.dviance.AwaPI.generic;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

/**
 * The global factory for the AwaPI.
 *
 */
public class GlobalFactory {

	public final static String API_PREFIX = "com.dviance.AwaPI.";
	
	private final static Map<String, Component> idPool = new HashMap<String, Component>();
	private final static Map<String, ArrayList<Component>> typePool = new HashMap<String, ArrayList<Component>>();
	
	/**
	 * Register a component into the global factory.
	 * If the component's id is already existing, the factory create a new instance and forget the old.
	 * Warning, the older instance is no more accessible in the factory but is always in memory !
	 * 
	 * @param id		Component's identifier (For example "avoider")
	 * @param type		Java package name (for example "components.navigation.Avoider")
	 * @param standardComponent		If your component is directly in the AwaPI, set true, else false.
	 * @return The component.
	 * @throws IllegalStateException
	 */
	public static Component addComponent(String id, String type, boolean standardComponent) throws IllegalStateException {
		
		Component obj = null;
		try {
			
			Class<?> cl = Class.forName( ((standardComponent) ? API_PREFIX : "") + type);
			Constructor<?> con = cl.getConstructor(String.class);
			obj = (Component)con.newInstance(id);
		} catch (NoSuchMethodException | SecurityException | InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException | ClassNotFoundException e) {
			
			throw new IllegalStateException("The class " + type + " doesn't exist");
		}
		
		idPool.put(id, obj);
		
		ArrayList<Component> l = typePool.get(type);
		if (l == null)
		{
			l = new ArrayList<>();
		}
		if (!l.contains(l))
			l.add(obj);
		
		typePool.put(type, l);
		
		return obj;
	}
	
	/**
	 * Get the list of all component of the same type.
	 * @param type		Component's type (for example "components.navigation")
	 * @return A list of component.
	 */
	public static ArrayList<Component> getComponentsByType(String type) {
		if (!typePool.containsKey(type)) {
			throw new IllegalStateException("No component typed " + type + "exists for now");
		}
		return typePool.get(type);
	}

	/**
	 * Get the component with the specified id.
	 * @param id 		Component's identifier (for example "avoider")
	 * @return The component or null if not found.
	 */
	public static Component getComponentById(String id) {
		if (!idPool.containsKey(id)) {
			throw new IllegalStateException("The component " + id + " doesn't exist");
		}
		return idPool.get(id);
	}
}
